Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -216,7 +230,7 @@ void DecodeNull(pb_istream_t* stream) {
   }
 }
 
-void Writer::WriteBool(bool bool_value) {
+Status Writer::WriteBool(bool bool_value) {
   return WriteVarint(bool_value);
```

Comment:
As an alternative to propagating status all the way up, consider storing it in a member variable inside `Writer` and making all write operations into no-ops if status is not ok, like the `Writer` [here](https://github.com/firebase/firebase-ios-sdk/blob/4c3fb51e32c419bc773d2b749fe392f21f29fb70/Firestore/core/src/firebase/firestore/local/leveldb_key.cc#L274-L282).

Reply:
I think that's a workaround due to the lack of StatusOr. @wilhuff?

Reply:
Oh, sorry for confusion, I linked `Reader` because `Writer` actually has no error handling. I mean something like:
```cpp
class Writer {
  // ...
  Status status_ = Status::OK();

  void WriteBool(bool value) {
    // No checking here because WriteVarint is a no-op
    // if Writer is in bad state
    WriteVarint(value);
  }

  void WriteVarint(int64_t value) {
    if (!status.ok()) {
      // If one of the previous operations failed, all subsequent
      // writes are no-ops
      return;
    }
    bool ok = pb_encode_varint(&stream_, value);
    if (!ok) {
      Fail(getInternalError(PB_GET_ERROR(&stream_)));
     // Or just Fail() and delegate the rest to Fail() method
    }
  }

  void Fail(const char* error) {
    status_ = Status{InternalError, error);
  }
  // Or alternatively
  void Fail() {
    status_ = Status{InternalError, PB_GET_ERROR(&stream_))};
  }
};
```

One upside is that functions calling several methods on `Writer` won't have to check/return after each operation.

Reply:
@rsgowman That's not a workaround for StatusOr. It allows writing simple [straight line reading logic](https://github.com/firebase/firebase-ios-sdk/blob/4c3fb51e32c419bc773d2b749fe392f21f29fb70/Firestore/core/src/firebase/firestore/local/leveldb_key.cc#L809).

I'm a fan of keeping the status as a member in the Writer, short-circuiting the innards, and pulling out out later at a single point.